### PR TITLE
add toolchain to java project created with `init`

### PR DESCRIPTION
**Pull request by reinsch82**
- resolves: #22682

Signed-off-by: Reinhold Degenfellner <rdegenfellner@gradle.com>

### CODE REVIEW

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle help

**Comment by reinsch82** [comment]
@bot-gradle test AFT

**Comment by reinsch82** [comment]
@bot-gradle clean

File: subprojects/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java [hunk 1]
```diff
@@ -183,44 +190,61 @@ public void setupProjectLayout() {
...
+            toolChainVersion);
+        initDescriptor.generate(settings);
+
+        initDescriptor.getFurtherReading(settings).ifPresent(link -> getLogger().lifecycle("Get more help with your project: {}", link));
+    }
+
+    java.util.Optional<JavaLanguageVersion> getJavaLanguageVersion(UserInputHandler inputHandler, Language language) {
+        if(isJvmLanguage(language)) {
+            JavaLanguageVersion current = JavaLanguageVersion.of(Jvm.current().getJavaVersion().getMajorVersion());
+            return of(inputHandler.selectOption("Select a Java toolchain to build the project", DefaultJavaLanguageVersion.getKnownVersions().collect(toList()), current));
```
Comment (tresat) [inline subprojects/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java:231]: ❗ I think the formatting of the version selection message needs to be changed.  With a long list, there should be more of a distinction between the option to select and the JDK version number - for the later version especially this is confusing.  They should perhaps say "JDK 11, JDK 12, etc."

image

File: subprojects/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java [hunk 2]
```diff
@@ -183,44 +190,61 @@ public void setupProjectLayout() {
...
+
+        boolean useIncubatingAPIs = shouldUseIncubatingAPIs(inputHandler);
+
+        BuildInitTestFramework testFramework = getBuildInitTestFramework(inputHandler, initDescriptor, modularizationOption);
+
+        String projectName = getProjectName(inputHandler, initDescriptor);
+
+        String packageName = getPackageName(inputHandler, initDescriptor, projectName);
+
+        java.util.Optional<JavaLanguageVersion> toolChainVersion = getJavaLanguageVersion(inputHandler, initDescriptor.getLanguage());
```
Comment (tresat) [inline subprojects/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java:207]: ❗ I think this functionality should initially only be added if the "Generate build using new APIs and behavior (some features may change in the next minor release)?" is selected.  This can be changed later, but people might be relying on the old behavior.

File: subprojects/build-init/src/main/java/org/gradle/buildinit/plugins/internal/BuildScriptBuilder.java [hunk 3]
```diff
@@ -513,11 +529,6 @@ public ChainedPropertyExpression(ExpressionValue left, ExpressionValue right) {
             this.right = right;
         }
 
-        @Override
-        public boolean isBooleanType() {
```
Comment (tresat) [inline subprojects/build-init/src/main/java/org/gradle/buildinit/plugins/internal/BuildScriptBuilder.java:517]: 💭 All the various small refactors like this default method and extracting methods for common functionality here really improve the code, but they make the core of this change harder to find when reviewing the PR.  Probably better to extract them to a separate PR next time, or if not possible, add some explanatory review comments to point people towards the more important stuff.

**Review by tresat — CHANGES_REQUESTED** [review]
I think supporting toolchains in the init plugin is great, but there are a few things here to address before this is ready.

@big-guy What do you think about the project dependency and making this opt-in to begin?

**Comment by big-guy** [comment]
I'm not sure about this. "Which toolchain should we use?" is the wrong question or at least not the most important.

We want to know which version of Java the library/application/whatever is targeting.  We can implement that as asking for a particular toolchain, but I don't think we should ask the user to pick a toolchain.

With that view, I don't think it matters what the user has installed. That's the entire point of toolchains.  So I think it would look something like this:
- Ask which Java runtime is going to be targeted
- If they don't answer... default to something reasonable like Java 11 or 17 or the current JVM or treat no answer as unspecified
- If they do answer, configure the build to target that runtime. The use of toolchains should flow from that.

WDYT?

File: subprojects/build-init/build.gradle.kts [hunk 4]
```diff
@@ -16,6 +16,7 @@ dependencies {
     implementation(project(":file-collections"))
     implementation(project(":dependency-management"))
     implementation(project(":platform-base"))
+    implementation(project(":platform-jvm"))
```
[the remark this answers is not in the document]
Comment (reinsch82) [reply subprojects/build-init/build.gradle.kts:19]: to my understanding `jvm-services` doesn't provide a `knownVersions` equivalent.
I also wouldn't consider it agnostic, it at least knows that jvm based projects exists.

File: subprojects/build-init/src/main/java/org/gradle/buildinit/plugins/internal/BuildScriptBuilder.java (hunk 3 above)
Comment (reinsch82) [reply subprojects/build-init/src/main/java/org/gradle/buildinit/plugins/internal/BuildScriptBuilder.java]: moved to
- #23374

File: subprojects/build-init/build.gradle.kts (hunk 4 above)
[the remark this answers is not in the document]
Comment (reinsch82) [reply subprojects/build-init/build.gradle.kts:19]: @big-guy WDYT

**Comment by reinsch82** [comment]
@big-guy 
We are calling it `toolchain` in the generated file. But I see your point.
Should we do anything to avoid confusion why the target version ends up as `toolchain` in the script?

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by eskatos** [comment]
We should also consider that there isn't auto-provisioning out of the box anymore.
Are we happy with that or do we want to apply the foojay-disco toolchain provisioning plugin right away?

**Comment by alllex** [comment]
> We should also consider that there isn't auto-provisioning out of the box anymore. Are we happy with that or do we want to apply the `foojay-disco` toolchain provisioning plugin right away?

@eskatos I would view this question from the user point of view. If I were as user, I would expect that whatever `init` produces should work out of the box, e.g. not break my IDE with "can't load project due to a missing JDK on the machine". With that in mind, I think the best way to be to make sure the generated project will just work regardless of user choices.

I would go with the following:
- If the requested toolchain / JDK for a target version is installed, then we simply reference it in the `java { toolchain { ... } }`
- Otherwise, we do one of the following:
  - (a) Inform the user the requested toolchain is not installed and whether they want it auto-installed
  - (b) Apply `foojay-disco` plugin without asking the user
  
My assumption is that in 99% of the cases the users will choose a target that already exists and all is good.

> If they don't answer... default to something reasonable like Java 11 or 17 or the current JVM or treat no answer as unspecified

@big-guy I would definitely not go with "unspecified" as we want to promote the usage of toolchains to make build more reproducible. Doing so is even more important in our own project generator tool. So I would strongly recommend defaulting to the latest LTS for the moment of Gradle release, which is Java 17 as of now.

**Comment by eskatos** [comment]
I agree that we should make it so that generated builds work out of the box.
My take would be to always add the foojay-disco plugin to that effect.
My rationale is that it would create less variations thus less surprise or support questions.
It would also make the build work on CI systems for which we can't detect in advance what's available.

File: subprojects/build-init/src/main/java/org/gradle/buildinit/plugins/internal/BuildScriptBuilder.java [hunk 5]
```diff
@@ -346,6 +347,16 @@ public BuildScriptBuilder block(@Nullable String comment, String methodName, Act
         return this;
     }
 
+    public BuildScriptBuilder javaLanguageVersion(JavaLanguageVersion languageVersion) {
```
Comment (eskatos) [inline subprojects/build-init/src/main/java/org/gradle/buildinit/plugins/internal/BuildScriptBuilder.java:350]: :lipstick: for explicitness: this is about configuring a java toolchain for a given language version
```suggestion
    public BuildScriptBuilder javaToolchainFor(JavaLanguageVersion languageVersion) {
```

File: subprojects/build-init/src/main/java/org/gradle/buildinit/plugins/internal/SimpleGlobalFilesBuildSettingsDescriptor.java [hunk 6]
```diff
@@ -53,6 +53,14 @@ private BuildScriptBuilder builder(InitSettings settings) {
...
             builder.methodInvocation(null, "include", settings.getSubprojects().toArray());
         }
+
+        if(settings.getJavaLanguageVersion().isPresent()){
+            builder.plugin(
+                "Apply the foojay plugin to allow automatic download of matching JDKs on other systems",
+                "org.gradle.toolchains.foojay-resolver",
+                "0.4.0");
+        }
+
```
Comment (eskatos) [inline subprojects/build-init/src/main/java/org/gradle/buildinit/plugins/internal/SimpleGlobalFilesBuildSettingsDescriptor.java:63]: :lipstick: I would put this after `pluginManagement {}` and before `rootProject.name` so that the code follows the order of the generated settings script more closely, for clarity

**Comment by reinsch82** [comment]
@big-guy 
It is generating the foojay plugin into the settings.gradle* automatically as @eskatos suggested.
I rephrased the question.
What we are not sure now ist whether the dependency to `platform-jvm` is reasonable.

File: subprojects/build-init/build.gradle.kts (hunk 4 above)
[the remark this answers is not in the document]
Comment (eskatos) [reply subprojects/build-init/build.gradle.kts:19]: FWIW `:build-init` already depends on a lot, including `:platform-native`. It also requires a full distribution for integ testing.

File: subprojects/build-init/src/integTest/groovy/org/gradle/buildinit/plugins/AbstractJvmLibraryInitIntegrationSpec.groovy [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+import org.gradle.integtests.fixtures.executer.ExecutionResult
+
+abstract class AbstractJvmLibraryInitIntegrationSpec extends AbstractInitIntegrationSpec{
+
+    @Override
+    String subprojectName() { 'lib' }
+
+    @Override
+    protected ExecutionResult run(String... tasks) {
```
Comment (big-guy) [inline subprojects/build-init/src/integTest/groovy/org/gradle/buildinit/plugins/AbstractJvmLibraryInitIntegrationSpec.groovy:27]: I don't think we should override what "run" means.  This is going to be a foot-gun in the future where run/succeeds/fails all behave differently WRT toolchains.

I think if you add a `setup` method, you can do `executer.withToolchainDetectionEnabled()` to have it always enabled or we can only enable this on tests that really require it.

File: subprojects/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java [hunk 8]
```diff
@@ -221,6 +233,42 @@ private static void validatePackageName(String packageName) {
         }
     }
 
+    static class JavaVersionSelection {
```
Comment (tresat) [inline subprojects/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java:236]: Instead of Java, it might be better for this to always ask about JDKs.  Saying Java is slightly confusing for people creating Groovy or Kotlin projects.

File: subprojects/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java [hunk 9]
```diff
@@ -221,6 +233,42 @@ private static void validatePackageName(String packageName) {
...
+
+    java.util.Optional<JavaLanguageVersion> getJavaLanguageVersion(boolean useIncubatingAPIs, UserInputHandler inputHandler, Language language) {
+        if (isJvmLanguage(language) && useIncubatingAPIs) {
+            JavaLanguageVersion current = JavaLanguageVersion.of(Jvm.current().getJavaVersion().getMajorVersion());
+            List<JavaVersionSelection> knownVersionSelections = DefaultJavaLanguageVersion.getKnownVersions().map(version -> new JavaVersionSelection(version)).collect(toList());
+            return of(inputHandler.selectOption(
+                "Which version of Java is this Project targeting",
+                knownVersionSelections,
+                new JavaVersionSelection(current)))
+                .map(JavaVersionSelection::getLanguageVersion);
```
Comment (big-guy) [inline subprojects/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java:261]: Do we need to constrain the version here at all? Can't someone provision a build with a higher Java version?

I would change this around to...
- Ask "Which version of Java is this project targeting?"
- Require this answer to be a number greater than whatever the minimum version our provisioning supports.

File: subprojects/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java [hunk 10]
```diff
@@ -221,6 +233,42 @@ private static void validatePackageName(String packageName) {
...
+            return "Java " + languageVersion;
+        }
+
+        public JavaLanguageVersion getLanguageVersion() {
+            return languageVersion;
+        }
+    }
+
+    java.util.Optional<JavaLanguageVersion> getJavaLanguageVersion(boolean useIncubatingAPIs, UserInputHandler inputHandler, Language language) {
+        if (isJvmLanguage(language) && useIncubatingAPIs) {
```
Comment (big-guy) [inline subprojects/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java:254]: I think the way we've done the other questions is to ask the init descriptor what it supports. I think we should do something similar here.

Comment (big-guy) [inline subprojects/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java:254]: Toolchain selection/provisioning is not considered incubating now, so I think it's OK to always ask this question vs looking at that flag.  The "use incubating API" flag was intended for using incubating features in the generated build and not introducing incubating behaviors to the init process.

**Review by tresat — APPROVED** [review]

File: subprojects/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java [hunk 11]
```diff
@@ -199,6 +206,10 @@ public void setupProjectLayout() {
 
         validatePackageName(packageName);
 
+        java.util.Optional<JavaLanguageVersion> toolChainVersion = getJavaLanguageVersion(useIncubatingAPIs, inputHandler, initDescriptor.getLanguage());
```
Comment (big-guy) [inline subprojects/build-init/src/main/java/org/gradle/buildinit/tasks/InitBuild.java:209]: I think all of the other options that can be prompted for also have a command-line equivalent. This option is similar to the testing framework, but even more specific to the Java ecosystem. 

The proper way of doing this would be to have a generic "target" option that could be reused for any ecosystem starting with the JVM, but that looks like more work to figure out how to do that in a nice way.

For now, I think I would not have a CLI option until someone complains.

**Review by big-guy — CHANGES_REQUESTED** [review]

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by big-guy** [comment]
@bot-gradle test and merge

**Comment by blindpirate** [comment]
@bot-gradle test and merge
